### ISSUE TITLE
[BUG] Playwright browser don't use system store with additional certificate authorities

### ISSUE THREAD

**Issue by aeris** [microsoft/playwright#28967]
Currently it seems Playwright don't consider additional system certificate authorities when starting a new browser and have no way to add manually one.

Normal browser, for example Firefox, notice any system store change even without a restart
Installing ca-cert authority with system package manager just show the certificate at the end on a running browser
![image](https://github.com/microsoft/playwright-python/assets/51246/9ab57ca9-3d35-4503-8f7a-02876bfa802c)

Playwright browser don't notice the additional authority even at start, and seem to use a static predefined list.

It leads to no way to test for website using authority outside the static list without skipping totally the TLS verification for all and any website and so remove every piece of security, as asked for such case [here](https://github.com/microsoft/playwright-python/issues/346).

### System info
- Playwright Version: v1.38.0
- Operating System: Arch Linux
- Browser: at least Chromium and Firefox

### Source code

```
$ cat test_ca.py
from playwright.sync_api import Page

def test_https_error_but_ca_is_system(page: Page):
    page.goto("https://cacert.org")
```

**Steps**

```
$ trust list | rg "CA Cert Signing Authority"
$ wget -4 https://cacert.org
--2024-01-12 15:39:30--  https://cacert.org/
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
ERROR: The certificate of 'cacert.org' is not trusted.

$ yay -S ca-certificates-cacert # Installing CACert certificate to system store

$ trust list | rg -C2 "CA Cert Signing Authority"
pkcs11:id=%16%B5%32%1B%D4%C7%F3%E0%E6%8E%F3%BD%D2%B0%3A%EE%B2%39%18%D1;type=cert
    type: certificate
    label: CA Cert Signing Authority
    trust: anchor
    category: authority
$ wget -4 https://cacert.org
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
HTTP request sent, awaiting response... 200 OK

$ pytest test_ca.py
```

[Describe expected behavior]

Test OK, using system certificate to access the site

[Describe actual behavior]

Test KO
```
E       playwright._impl._api_types.Error: SEC_ERROR_UNKNOWN_ISSUER
E       =========================== logs ===========================
E       navigating to "https://cacert.org/", waiting until "load"
E       ============================================================
```

**Comment by mxschmitt** [comment]
We are unfortunately not familiar with Arch Linux, its also not a Linux distribution [we support](https://playwright.dev/docs/intro#system-requirements) - have you tried it on a [supported Linux distribution](https://playwright.dev/docs/intro#system-requirements)?

I found this which looks related, have you tried that? https://superuser.com/questions/1717914/make-chrome-trust-the-linux-system-certificate-store-or-select-certificates-via

(Playwright's Chromium is the same as a normal Chromium, we don't do any certificate related changes. Actually no changes at all on the Chromium side as of today.)

**Comment by aeris** [comment]
Hello
I got the same trouble on Node LTS + Debian 12
Minimal reproducible docker image available here
https://github.com/aeris/playwrigth-system-ca

**Comment by aeris** [comment]
For the "solution" on the superuser forum, it's not possible for playwright, certificate store only exists in browser profile, and so is just clean each time you restart playwright (no persistence)

And I can't find a reliable and portable way to get the running profile directory from inside playwright execution to be able to inject a new ca certificate.

I don't also know the difference between playwright browsers and standard user browsers, but there is a different behavior.
Everyday browsers seem using system certificates, even noticing change without a restart, but the ones playwright start seem to be totally isolated and don't notice system store change.

**Comment by mxschmitt** [comment]
For Chromium it seems doable via:

```Dockerfile
# For Chromium
# https://chromium.googlesource.com/chromium/src/+/master/docs/linux/cert_management.md
RUN apt install libnss3-tools
RUN mkdir -p $HOME/.pki/nssdb
RUN certutil -d sql:$HOME/.pki/nssdb -A -t "C,," -n badssl-com -i badssl-com.pem
```

For Firefox it wasn't working for me, since as you said we are using in-memory browser profiles which end up creating temp browser profiles. My attempt on following [this](https://wiki.mozilla.org/CA/AddRootToFirefox) didn't work.

**Comment by Dzeneralen** [comment]
Would be nice to have a way to either supply `policies.json` to be able to add certificates to Firefox or some other workaround. For now disabling SSL errors seems like the only solution.  Similar discussion in https://github.com/microsoft/playwright/issues/18115

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your contribution to our project. This issue has been closed due to its limited upvotes and recent activity, and insufficient feedback for us to effectively act upon. Our priority is to focus on bugs that reflect higher user engagement and have actionable feedback, to ensure our bug database stays manageable.

Should you feel this closure was in error, please create a new issue and reference this one. We're open to revisiting it given increased support or additional clarity. Your understanding and cooperation are greatly appreciated.

**Comment by alexaka1** [comment]
Can we reopen this?

**Comment by fwdekker** [comment]
@alexaka1 For Firefox, you can set env variable `PLAYWRIGHT_FIREFOX_POLICIES_JSON=<file>` to load a `policies.json` file from the specified `<file>`. You can [use `policies.json` to specify custom certificates that Firefox should trust](https://mozilla.github.io/policy-templates/#certificates--install). If it doesn't seem to work, make sure you check `about:policies` and see if there's an error displayed in the error tab on that page. I had some trouble because I was loading a `.pem` file which also contained a private key, but Firefox requires the file to contain only the public key.

For Chromium, I suppose [the solution by mxschmitt](https://github.com/microsoft/playwright/issues/28967#issuecomment-1899384015) is your best bet.

**Comment by macgeneral** [comment]
For anyone who might is interested in working around this bug/behaviour.

The following snippet fixes it for Chromium, Firefox and Webkit (verified with the -noble images of Playwright 1.58.0 to 1.58.2).

```Dockerfile
FROM mcr.microsoft.com/playwright:${PLAYWRIGHT_VERSION:?}

USER root

// Do your ca-certificates stuff here

RUN apt-get update -qq && \
    apt-get install -y --no-install-recommends p11-kit p11-kit-modules && \
    rm -rf /var/lib/apt/lists/* && \
    # Replace every libnssckbi.so (system + any bundled copy) with p11-kit-trust.so
    # so Chromium uses the system CA trust store including the custom CAs.
    # Paths are resolved dynamically to handle aarch64 vs x86_64.
    P11=$(find /usr/lib -name 'p11-kit-trust.so' | head -1) && \
    find /usr -name 'libnssckbi.so' -exec ln -sf "$P11" {} \; && \
    find /ms-playwright -name 'libnssckbi.so' 2>/dev/null -exec ln -sf "$P11" {} \; && \
    # Firefox enterprise policy: install each corporate CA individually.
    # Playwright reads the policy path from the PLAYWRIGHT_FIREFOX_POLICIES_JSON
    # env var (see playwright.cfg: browser.policies.alternatePath).
    # Use a fixed absolute path so the env var is reliable.
    mkdir -p /etc/playwright && \
    certs=$(find /usr/local/share/ca-certificates -name '*.crt' | \
      awk '{printf "%s\"%s\"", (NR>1?",":""), $0} END{print ""}') && \
    printf '{"policies":{"Certificates":{"Install":[%s]}}}\n' "$certs" \
      > /etc/playwright/firefox-policies.json && \
    chmod a+r /etc/playwright/firefox-policies.json

# Tell Playwright's Firefox to load our policies file.
ENV PLAYWRIGHT_FIREFOX_POLICIES_JSON=/etc/playwright/firefox-policies.json

USER pwuser
```

**Comment by markedwards** [comment]
Also facing this issue using playwright installed via pnpm.  The solution proposed by @macgeneral above works for Chromium, but could not make it work for WebKit or Firefox in this context.

Can this issue be re-opened?  Its a pretty big blind spot in Playwright to not be able to trust a custom CA for the purpose of mocking.

**Comment by alexaka1** [comment]
The only reliable solution I have found is in CI you have to prewarm the browser (ie screenshot example.com) and only after can you add your custom CAs. I use `mkcert`, this could be an mkcert quirk, I dunno. I've actually ran into a different bug since then, but I have stopped caring to fix it.

Even if your browser trust your CA, Node does not. And playwright pipes some calls through node, ie page.route(), which can still fail at your CA even if you have the new trust system CA option enabled. Edit: Okay this may be specific to dotnet dev-certs, but this is a much less annoying bug overall.

**Comment by fwdekker** [comment]
In Firefox, you can preload certificates with `policies.json`, see specifically [Certificates -> Install](https://mozilla.github.io/policy-templates/#certificates--install).

In Chromium, I was able to reverse engineer the profile structure. You can do the following steps even before starting Chromium for the first time; Chromium will keep the file intact.
1. Create a SQLite database in the `<profile>/Default/` directory called `ServerCertificate`.
2. Create the required two tables:
   ```sql
   CREATE TABLE certificates(sha256hash_hex TEXT PRIMARY KEY,der_cert BLOB NOT NULL,trust_settings BLOB NOT NULL)
   CREATE TABLE meta(key LONGVARCHAR NOT NULL UNIQUE PRIMARY KEY, value LONGVARCHAR)
   ```
3. Populate the `meta` table (not sure if technically required?):
   ```sql
   INSERT INTO "meta" ("key", "value") VALUES ('mmap_status', '-1');
   INSERT INTO "meta" ("key", "value") VALUES ('version', '1');
   INSERT INTO "meta" ("key", "value") VALUES ('last_compatible_version', '1');
   ```
4. Prepare data you have to insert into the `certificates` table:
   * Find the public key of the certificate you want to insert. If you have a `.pem` file, it's the part between `-----BEGIN CERTIFICATE-----` and `-----END CERTIFICATE-----`, with all lines concatenated into one long line. So no `\n` chars. Store this as a textual string.
   * Calculate the sha256 hash of the above string.
5. Insert the row:
   ```sql
   INSERT INTO "certificates" ("sha256hash_hex", "der_cert", "trust_settings") VALUES ('<sha256sum>', '<public key>', X'0a020803');
   ```
   Keep the `X'0a020803'` as is. Don't change it. I don't know what it means.

Safety not guaranteed.

**Comment by wbolster** [comment]
there is no need for manual sqlite operations. use `update-ca-certificates` (debian/ubuntu, including playwright container images) and `certutil`.

a shell script like this works for the [playwright container images](https://playwright.dev/docs/docker) running in a CI environment like gitlab CI. the system store works for tools like curl and openssl. the certutil stuff works for playwright running chrome/chromium.

```bash
#!/usr/bin/env bash

# Add a trusted certificate authority (CA).

set -euo pipefail

main() {
    local file="${1:?no certificate specified}"

    # System store (curl, openssl, …)
    install -m 644 "$file" "/usr/local/share/ca-certificates/$(basename "${file%.crt}").crt"
    update-ca-certificates

    # NSS (Chrome, Firefox, …)
    local nssdb="$HOME/.local/share/pki/nssdb"
    if [[ ! -d "$nssdb" ]]; then
        mkdir -p "$nssdb"
        certutil --empty-password -d "$nssdb" -N
    fi
    certutil -d "sql:$nssdb" -A -t C,, -n "$(basename "$file")" -i "$file"
    certutil -d "sql:$nssdb" -L
}

main "$@"
```
